mpeg2/src/temi.rs

```diff
@@ -0,0 +1,282 @@
+#[allow(non_camel_case_types)]
+#[repr(C)]
+pub enum TimestampLength {
+    None = 0,
+    ThirtyTwoBits = 1,
+    SixtyFourBits = 2,
+}
+
+impl Default for TimestampLength {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+#[derive(Debug, Default, PartialEq)]
+pub struct TEMITimelineDescriptor {
+    pub has_timestamp: TimestampLength,
+    pub timescale: u32,
+    pub media_timestamp: u64,
+
+    pub ntp_timestamp: Option<u64>,
+    pub ptp_timestamp: Option<u128>,
+
+    pub drop: bool,
+    pub frames_per_tc_seconds: u16,
+    pub duration: u16,
+    pub has_timecode: TimestampLength,
+    pub time_code: u64,
+
+    pub force_reload: bool,
+    pub paused: bool,
+    pub discontinuity: bool,
+    pub timeline_id: u8,
+}
+
+impl TEMITimelineDescriptor {
+    pub fn encode_len(&self) -> usize {
+        let mut len = if self.has_timestamp == TimestampLength::ThirtyTwoBits {
+            13
+        } else if self.has_timestamp == TimestampLength::SixtyFourBits {
+            17
+        } else {
+            5
+        };
+        if self.ntp_timestamp.is_some() {
+            len += 8;
+        }
+        if self.ptp_timestamp.is_some() {
+            len += 10;
+        }
+
+        len += if self.has_timecode == TimestampLength::ThirtyTwoBits {
+            7
+        } else if self.has_timecode == TimestampLength::SixtyFourBits {
+            12
+        } else {
+            0
+        };
+        len
+    }
+
+    pub fn encode<W: Write>(&self, mut w: W) -> Result<usize, EncodeError> {
+        let mut buf = vec![0u8; self.encode_len()];
+        buf[0] = AF_DESCR_TAG_TIMELINE;
+        buf[2] = (self.has_timestamp as u8) << 6;
+
+        if self.force_reload {
+            buf[2] |= 0b0000_0010;
+        }
+        if self.paused {
+            buf[2] |= 0b0000_0001;
+        }
+        if self.discontinuity {
+            buf[3] = 0b1000_0000;
+        }
+        buf[4] = self.timeline_id;
+
+        let mut ret = 5;
+        if self.has_timestamp != TimestampLength::None {
+            BigEndian::write_u32(&mut buf[5..9], self.timescale);
+            if self.has_timestamp == TimestampLength::ThirtyTwoBits {
+                BigEndian::write_u32(&mut buf[9..13], self.media_timestamp as u32);
+                ret += 8;
+            } else if self.has_timestamp == TimestampLength::SixtyFourBits {
+                BigEndian::write_u64(&mut buf[9..17], self.media_timestamp);
+                ret += 12;
+            }
+        }
+
+        if let Some(ntp_ts) = self.ntp_timestamp {
+            buf[2] |= 0b0010_0000;
+            BigEndian::write_u64(&mut buf[ret..(ret + 8)], ntp_ts);
+            ret += 8;
+        }
+        if let Some(ptp_ts) = self.ptp_timestamp {
+            buf[2] |= 0b0001_0000;
+            BigEndian::write_u16(&mut buf[ret..(ret + 2)], (ptp_ts >> 64) as u16);
+            ret += 2;
+            BigEndian::write_u64(&mut buf[ret..(ret + 8)], ptp_ts as u64);
```

Comment:
These write ranges don't need to have an end specified. I would recommend just removing them:

```suggestion
            BigEndian::write_u64(&mut buf[ret..], ptp_ts as u64);
```

Makes the code easier to read and check.

Reply:
will remove all the redundant end indexes.